cmd/store.go

```diff
@@ -31,64 +35,132 @@ func (s *store) Get(IP string) (string, error) {
+func (s *store) RemoveRoleFromNamespace(namespace string, role string) {
+	// Make sure to remove the full ARN of roles to ensure string matching works
+	roleARN := s.iam.roleARN(role)
 
-	if oldPod.Status.PodIP != newPod.Status.PodIP {
-		s.OnDelete(oldPod)
-		s.OnAdd(newPod)
+	ar := s.rolesByNamespace[namespace]
+	for i := range ar {
+		if ar[i] == roleARN {
+			ar = append(ar[:i], ar[i+1:]...)
+			break
+		}
 	}
+	s.mutex.Lock()
+	s.rolesByNamespace[namespace] = ar
+	s.mutex.Unlock()
+}
+
+// DeleteNamespace removes all role mappings from a namespace
+func (s *store) DeleteNamespace(namespace string) {
+	s.mutex.Lock()
+	delete(s.rolesByNamespace, namespace)
+	s.mutex.Unlock()
 }
 
-// OnDelete is called when a pod is deleted.
-func (s *store) OnDelete(obj interface{}) {
-	pod, ok := obj.(*api.Pod)
-	if !ok {
-		deletedObj, dok := obj.(kcache.DeletedFinalStateUnknown)
-		if dok {
-			pod, ok = deletedObj.Obj.(*api.Pod)
+// checkRoleForNamespace checks the 'database' for a role allowed in a namespace,
+// returns true if the role is found, otheriwse false
+func (s *store) checkRoleForNamespace(role string, namespace string) bool {
+	ar := s.rolesByNamespace[namespace]
+	if ar == nil {
+		log.Warnf("Role:%s on namespace:%s not found.", role, namespace)
+		return false
+	}
+	for i := range ar {
+		if ar[i] == role {
+			log.Debugf("Role:%s on namespace:%s found.", role, namespace)
+			return true
 		}
 	}
+	log.Warnf("Role:%s on namespace:%s not found.", role, namespace)
+	return false
+}
 
-	if !ok {
-		log.Errorf("Expected Pod but OnDelete handler received %+v", obj)
-		return
+func (s *store) CheckNamespaceRestriction(role string, ip string) bool {
+	// if the namespace restrictions are not in place early out true
+	if !s.namespaceRestriction {
+		return true
 	}
 
-	if pod.Status.PodIP != "" {
-		s.mutex.Lock()
-		delete(s.rolesByIP, pod.Status.PodIP)
-		s.mutex.Unlock()
+	// if the role is the default role you are also good
+	if role == s.defaultRole {
+		return true
 	}
+
+	ns := s.namespaceByIP[ip]
+
+	return s.checkRoleForNamespace(role, ns)
+}
+
+func (s *store) DumpRolesByIP() map[string]string {
+	return s.rolesByIP
+}
+
+func (s *store) DumpRolesByNamespace() map[string][]string {
+	return s.rolesByNamespace
+}
+
+func (s *store) DumpNamespaceByIP() map[string]string {
+	return s.namespaceByIP
 }
 
-func newStore(key string, defaultRole string) *store {
+func newStore(key string, defaultRole string, namespaceRestriction bool, namespaceKey string, iamInstance *iam) *store {
 	return &store{
-		defaultRole: defaultRole,
-		iamRoleKey:  key,
-		rolesByIP:   make(map[string]string),
+		defaultRole:          defaultRole,
+		iamRoleKey:           key,
+		namespaceKey:         namespaceKey,
+		namespaceRestriction: namespaceRestriction,
+		rolesByIP:            make(map[string]string),
+		rolesByNamespace:     make(map[string][]string),
+		namespaceByIP:        make(map[string]string),
```

Comment:
Yeah there's another PR that was proposing that. But it would need serious rebasing and does not include namespaces obviously. Not sure why they don't use that in the kubernetes code base for DNS though and if there is any trade-off.